include/mpark/config.hpp

```diff
@@ -50,7 +50,7 @@
 #define MPARK_BUILTIN_UNREACHABLE
 #endif
 
-#if __has_builtin(__type_pack_element)
+#if __has_builtin(__type_pack_element) && !(defined(__ICC))
```

Comment:
Following this quick check on Linux, I think your problem is most likely macOS specific:
  https://godbolt.org/z/6v5GMxz4v

I would this recommend to limit this patch to ICC + Apple:
```suggestion
+#if __has_builtin(__type_pack_element) && !(defined(__ICC) && defined(__APPLE__))
```

Reply:
Responded [here](https://github.com/spack/spack/pull/22727#discussion_r605978256):

> I originally thought this was an Apple specific issue, but I think it is just dependent on the toolchain icc is using (I may be saying this wrong; I really don't understand how Intel uses the underlying toolchain's preprocessor).  The reason this is passing on Linux is that with the gcc10.2 toolchain, `__has_builtin(__type_pack_element)` is evaluating to false: https://godbolt.org/z/zeevevqcW
>
> My concern is that if in the future GCC adds `__type_pack_element` support, then Intel with that tool-chain will start failing.  It may also fail on linux if someone compiles using an LLVM toolchain: [intel doc on setting clang](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/compatibility-options/clangxx-name.html.)  Other than compile time, I don't think there is much risk in always using mpark's implementation of type_pack_element.  
>
> But I am open to maybe making it `__clang__` specific for now and dealing with it in the future when it arises :) 

Ok, then let's keep this for ICC in general.